### PR TITLE
separate xds cache per type and update cache key

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
**Please provide a description of this PR:**

The main intention is to reduce lock contention.

### CODE REVIEW

**Comment by hzxuzhonghu** [comment]
@howardjohn @ramaraochavali

File: pilot/pkg/model/xds_cache.go [hunk 1]
```diff
@@ -71,79 +71,70 @@ func size(cs int) {
...
-type XdsCache interface {
+type xdsCache[K comparable] interface {
 	// Run starts a background thread periodically flush evicted indexes
 	Run(stop <-chan struct{})
 	// Add adds the given XdsCacheEntry with the value for the given pushContext to the cache.
 	// If the cache has been updated to a newer push context, the write will be dropped silently.
 	// This ensures stale data does not overwrite fresh data when dealing with concurrent
 	// writers.
-	Add(entry XdsCacheEntry, pushRequest *PushRequest, value *discovery.Resource, opts ...CacheOption)
+	Add(key K, entry dependents, pushRequest *PushRequest, value *discovery.Resource)
```
Comment (ramaraochavali) [inline pilot/pkg/model/xds_cache.go:89]: nit: doc comments are incorrect now

File: pilot/pkg/model/generic_xds_cache.go [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+type XdsCacheImpl struct {
+	cds xdsCache[uint64]
+	eds xdsCache[uint64]
+	rds xdsCache[uint64]
+	sds xdsCache[string]
+}
+
+// GenericXdsCache interface defines a store for caching XDS responses.
+// All operations are thread safe.
+type GenericXdsCache interface {
```
Comment (ramaraochavali) [inline pilot/pkg/model/generic_xds_cache.go:34]: keeping the name of this as "XdsCache" and renaming the interface in xds_cache as "typedXdsCache" may be more readable as that is per type?

File: pilot/pkg/model/generic_xds_cache.go [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+
+const (
+	CDSType = "cds"
+	EDSType = "eds"
+	RDSType = "rds"
+	SDSType = "sds"
+)
+
+// NewGenericXdsCache returns an instance of a cache.
+func NewGenericXdsCache() GenericXdsCache {
```
Comment (ramaraochavali) [inline pilot/pkg/model/generic_xds_cache.go:78]: Similarly rename this to XdsCache?

File: pilot/pkg/model/generic_xds_cache.go [hunk 4]
```diff
@@ -0,0 +1,264 @@
+// Copyright Istio Authors
```
Comment (ramaraochavali) [inline pilot/pkg/model/generic_xds_cache.go:1]: May be retain the existing functionality in xds_cache and for new per type xds cache, move it to typed_xds_cache or xds_cache_type?

**Review by ramaraochavali — COMMENTED** [review]
Added few comments

File: pilot/pkg/model/generic_xds_cache.go (hunk 2 above)
Comment (hzxuzhonghu) [reply pilot/pkg/model/generic_xds_cache.go:34]: SG, i added `GenericXdsCache` first before renaming the original XdsCache

File: pilot/pkg/xds/sds.go [hunk 5]
```diff
@@ -51,8 +52,14 @@ type SecretResource struct {
...
 
-func (sr SecretResource) Key() string {
-	return sr.SecretResource.Key() + "/" + sr.pkpConfHash
+func (sr SecretResource) Type() string {
+	return model.SDSType
+}
+
+func (sr SecretResource) Key() any {
+	h := hash.New()
+	h.Write([]byte(sr.SecretResource.Key() + "/" + sr.pkpConfHash))
```
Comment (howardjohn) [inline pilot/pkg/xds/sds.go:61]: I thought the whole point of this change was to NOT hash the SDS type? We cannot safely hash because a collision will be a CVE

File: pilot/pkg/networking/core/v1alpha3/route/route_cache.go [hunk 6]
```diff
@@ -177,11 +180,19 @@ func (r *Cache) Key() string {
...
-	return h.Sum()
+	return h.Sum64()
 }
 
 func hashToBytes(number model.ConfigHash) []byte {
-	big := new(big.Int)
-	big.SetUint64(uint64(number))
-	return big.Bytes()
+	v := uint64(number)
+	return []byte{
```
[the remark this answers is not in the document]
Comment (hzxuzhonghu) [reply pilot/pkg/networking/core/v1alpha3/route/route_cache.go:188]: for perf

File: pilot/pkg/xds/sds.go (hunk 5 above)
Comment (hzxuzhonghu) [reply pilot/pkg/xds/sds.go:61]: Ok, i can revert the original impl

File: pilot/pkg/model/xds_cache.go [hunk 7]
```diff
@@ -104,329 +47,218 @@ type XdsCache interface {
...
-				l.evictQueue = l.evictQueue[:0:1000]
-				l.recordDependentConfigSize()
-				l.mu.Unlock()
-			case <-stop:
-				return
-			}
-		}
-	}()
-}
+const (
```
Comment (ramaraochavali) [inline pilot/pkg/model/xds_cache.go:70]: May be we should use typeUrls for safety https://github.com/istio/istio/blob/master/pilot/pkg/xds/v3/model.go#L26

File: pilot/pkg/xds/eds.go [hunk 8]
```diff
@@ -434,17 +434,6 @@ func canSendPartialFullPushes(req *model.PushRequest) bool {
 	return true
 }
 
-func canUseOldEdsCache(req *model.PushRequest) bool {
```
Comment (ramaraochavali) [inline pilot/pkg/xds/eds.go:437]: Is this removal intentional? How are we handling PA changes now?

File: pilot/pkg/model/xds_cache.go (hunk 7 above)
Comment (hzxuzhonghu) [reply pilot/pkg/model/xds_cache.go:70]: it is just a flag, not being stored at all

File: pilot/pkg/xds/eds.go (hunk 8 above)
Comment (hzxuzhonghu) [reply pilot/pkg/xds/eds.go:437]: Yes, i have moved this handling to 
```
func (x XdsCacheImpl) Clear(s sets.Set[ConfigKey]) {
	x.cds.Clear(s)
	// clear all EDS cache for PA change
	if HasConfigsOfKind(s, kind.PeerAuthentication) {
		x.eds.ClearAll()
	} else {
		x.eds.Clear(s)
	}
	x.rds.Clear(s)
	x.sds.Clear(s)
}
```

Comment (ramaraochavali) [reply pilot/pkg/xds/eds.go:437]: Got it

File: pilot/pkg/model/xds_cache.go (hunk 7 above)
Comment (ramaraochavali) [reply pilot/pkg/model/xds_cache.go:70]: I know. I am just suggesting use the same const

**Review by ramaraochavali — COMMENTED** [review]
LGTM. Will leave it open for @howardjohn to take a look

File: pilot/pkg/model/xds_cache.go [hunk 9]
```diff
@@ -104,329 +47,218 @@ type XdsCache interface {
...
-	for ckey := range configs {
-		referenced := l.configIndex[ckey.HashCode()]
-		delete(l.configIndex, ckey.HashCode())
-		for key := range referenced {
-			l.store.Remove(key)
-		}
+func (x XdsCacheImpl) Clear(s sets.Set[ConfigKey]) {
+	x.cds.Clear(s)
+	// clear all EDS cache for PA change
+	if HasConfigsOfKind(s, kind.PeerAuthentication) {
```
Comment (howardjohn) [inline pilot/pkg/model/xds_cache.go:156]: should we remove the GetWithRequest now? (can be a followup)

**Review by howardjohn — APPROVED** [review]
